### PR TITLE
client: fix for jetty session timeout

### PR DESCRIPTION
**Pull request by marcaurele**
## Description

Related to https://github.com/apache/cloudstack/pull/2329#issuecomment-547904987 the session timeout is not taken into account. This change creates a session handler explicitly and the timeout value should be correctly taken into account.

## Types of changes
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

This has not been tested yet.

### CODE REVIEW

**Comment by andrijapanicsb** [comment]
thx! let me test that

**Comment by andrijapanicsb** [comment]
@blueorangutan package

**Comment by andrijapanicsb** [comment]
@blueorangutan test matrix

**Comment by marcaurele** [comment]
I did some testing on a simple server and I had to set the timeout after the server was started. I pushed an update, let me know if it's working now @andrijapanicsb

**Comment by andrijapanicsb** [comment]
thx @marcaurele.

@blueorangutan package

**Comment by andrijapanicsb** [comment]
@marcaurele @nvazquez - I've tested both your solutions - it works fine! (tried both setting session to 2minutes and 45 minutes) - so all good!

File: client/src/main/java/org/apache/cloudstack/ServerDaemon.java [hunk 1]
```diff
@@ -24,6 +24,7 @@
 import java.net.URL;
 import java.util.Properties;
 
+import com.cloud.utils.Pair;
```
Comment (nvazquez) [inline client/src/main/java/org/apache/cloudstack/ServerDaemon.java]:
```suggestion
```

File: client/src/main/java/org/apache/cloudstack/ServerDaemon.java [hunk 2]
```diff
@@ -175,7 +177,8 @@ public void start() throws Exception {
         createHttpConnector(httpConfig);
 
         // Setup handlers
-        server.setHandler(createHandlers());
+        Pair<SessionHandler,HandlerCollection> pair = createHandlers();
```
Comment (nvazquez) [inline client/src/main/java/org/apache/cloudstack/ServerDaemon.java]:
```suggestion
```

File: client/src/main/java/org/apache/cloudstack/ServerDaemon.java [hunk 3]
```diff
@@ -175,7 +177,8 @@ public void start() throws Exception {
...
+        server.setHandler(pair.second());
```
Comment (nvazquez) [inline client/src/main/java/org/apache/cloudstack/ServerDaemon.java]:
```suggestion
        server.setHandler(createHandlers());
```

File: client/src/main/java/org/apache/cloudstack/ServerDaemon.java [hunk 4]
```diff
@@ -236,11 +241,10 @@ private void createHttpsConnector(final HttpConfiguration httpConfig) {
         }
     }
 
-    private HandlerCollection createHandlers() {
+    private Pair<SessionHandler,HandlerCollection> createHandlers() {
```
Comment (nvazquez) [inline client/src/main/java/org/apache/cloudstack/ServerDaemon.java]:
```suggestion
    private HandlerCollection createHandlers() {
```

File: client/src/main/java/org/apache/cloudstack/ServerDaemon.java [hunk 5]
```diff
@@ -259,14 +263,14 @@ private HandlerCollection createHandlers() {
...
-        // Redirect root context handler
+        // Redirect root context handler_war
         MovedContextHandler rootRedirect = new MovedContextHandler();
         rootRedirect.setContextPath("/");
         rootRedirect.setNewContextURL(contextPath);
         rootRedirect.setPermanent(true);
 
         // Put rootRedirect at the end!
-        return new HandlerCollection(log, gzipHandler, rootRedirect);
+        return new Pair<>(webApp.getSessionHandler(), new HandlerCollection(log, gzipHandler, rootRedirect));
```
Comment (nvazquez) [inline client/src/main/java/org/apache/cloudstack/ServerDaemon.java]:
```suggestion
        return new HandlerCollection(log, gzipHandler, rootRedirect);
```

File: client/src/main/java/org/apache/cloudstack/ServerDaemon.java [hunk 6]
```diff
@@ -184,6 +187,8 @@ public void start() throws Exception {
         createHttpsConnector(httpConfig);
 
         server.start();
+        // Must set the session timeout after the server has started
+        pair.first().setMaxInactiveInterval(sessionTimeout * 60);
```
Comment (nvazquez) [inline client/src/main/java/org/apache/cloudstack/ServerDaemon.java]:
```suggestion
        sessionHandler.setMaxInactiveInterval(sessionTimeout * 60);
```

**Comment by nvazquez** [comment]
@marcaurele please check PR #3663

**Comment by marcaurele** [comment]
@nvazquez IMO it does not look well coded to keep a reference of the `SessionHandler` as a class member to only skip having to return this pair of objects.
Then, I don't understand why you open a new PR to perform those changes? Is that so urgent? Aren't your comment enough?

**Comment by nvazquez** [comment]
@marcaurele no urgency, was trying to propose an improvement and tried with the comments but thought those were not being added until I click Submit on the review. So I created the PR for clarity but didn't realise the comments were already added, please discard them. Thanks for the fix

**Comment by andrijapanicsb** [comment]
Guys, please let me know once you have agreed on the way forward, so I can merge this one.

**Comment by nvazquez** [comment]
I have closed the other PR, I'm ok with this one

**Comment by marcaurele** [comment]
@nvazquez I wish we could return multiple argument without creating a custom class in Java. Thanks for taking my point in consideration.

**Comment by marcaurele** [comment]
Rebase done to only leave the *good* commit. Good for review & merge on my side.

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test matrix

**Comment by andrijapanicsb** [comment]
the latest code is NOT ok.
I've set timeout to 2 minutes and I don't get logged out on 2m30sec (nor on 3m30sec nor on 4m15sec - tested just in case...)

**Comment by rhtyd** [comment]
@marcaurele can you try https://stackoverflow.com/a/56091476

**Comment by marcaurele** [comment]
@andrijapanicsb weird as I don't see any difference so far from the previous 2 commits. Can you double check you're using the right branch and cleared up the compiled classes of the client module, just in case.

@rhtyd that's what it's doing, but I found out that the session must be set after doing the `server.start()`. I modified the jetty example embbeded project to try out.

**Comment by andrijapanicsb** [comment]
my bad, give me 2 minutes...

**Comment by andrijapanicsb** [comment]
all good @marcaurele.

LGTM from my side (manual testing).

**Comment by andrijapanicsb** [comment]
merging based on 2 LGTM, manual testing and automated test results.

thx @marcaurele !

**Comment by marcaurele** [comment]
For the record, this is the small embedded app to validate: https://github.com/marcaurele/embedded-jetty-uber-jar/tree/session-test

**Pull request merged**